v0/destinations/salesforce/utils.js

```diff
@@ -0,0 +1,146 @@
+    } else {
+      SF_TOKEN_URL = SF_TOKEN_REQUEST_URL;
+    }
+    const authUrl = `${SF_TOKEN_URL}?username=${
+      destination.Config.userName
+    }&password=${encodeURIComponent(
+      destination.Config.password
+    )}${encodeURIComponent(destination.Config.initialAccessToken)}&client_id=${
+      destination.Config.consumerKey
+    }&client_secret=${destination.Config.consumerSecret}&grant_type=password`;
+    const salesforceAuthorisationData = await httpPOST(authUrl, {});
+    const salesforceAxiosResponse = processAxiosResponse(
+      salesforceAuthorisationData
+    );
+    // If the request fails, throwing error.
+    if (salesforceAuthorisationData.success === false) {
+      const { error } = salesforceAuthorisationData.response.response.data;
+      throw new ApiError(
+        `Request Failed for Salesforce, access token could not be generated due to ${error}`,
+        400,
+        {
+          scope: TRANSFORMER_METRIC.MEASUREMENT_TYPE.API.SCOPE,
+          meta: TRANSFORMER_METRIC.MEASUREMENT_TYPE.API.META.RETRYABLE
+        },
+        salesforceAxiosResponse.response,
+        undefined,
+        DESTINATION
+      );
+    }
+    const token = salesforceAuthorisationData.response.data;
+    return {
+      token: `Bearer ${token.access_token}`,
+      instanceUrl: token.instance_url
+    };
+  });
+};
+
+const processResponseHandler = (
+  destResponse,
+  sourceMessage,
+  stage,
+  authKey
+) => {
+  const { status, response } = destResponse;
+
+  // if the response from destination is not a success case build an explicit error
+  if (!isHttpStatusSuccess(status)) {
+    if (status === 401 && authKey) {
+      if (
+        response &&
+        Array.isArray(response) &&
+        response[0].errorCode === "INVALID_SESSION_ID"
+      ) {
+        // checking for invalid/expired token errors and evicting cache in that case
+        // rudderJobMetadata contains some destination info which is being used to evict the cache
+        ACCESS_TOKEN_CACHE.del(authKey);
+        throw new ApiError(
+          `Request Failed for Salesforce due to ${response[0].message}, (Retryable).${sourceMessage}`,
+          500,
+          {
+            scope: TRANSFORMER_METRIC.MEASUREMENT_TYPE.API.SCOPE,
+            meta: TRANSFORMER_METRIC.MEASUREMENT_TYPE.API.META.RETRYABLE
+          },
+          sourceMessage,
+          undefined,
+          DESTINATION
+        );
+      }
+    }
+    // check the error message
+    let errorMessage = "";
+    if (response && Array.isArray(response)) {
+      errorMessage = response[0].message;
+    }
+
+    throw new ApiError(
+      `Request Failed for ${DESTINATION}: ${status} due to ${errorMessage}, Aborted.`,
+      status,
+      {
+        scope: TRANSFORMER_METRIC.MEASUREMENT_TYPE.API.SCOPE,
+        meta: getDynamicMeta(status)
+      },
+      400,
+      undefined,
+      DESTINATION
+    );
+  }
+  if (isHttpStatusSuccess(status)) {
```

Comment:
Why are we doing this again? We are already evicting the cache in the case of `401` above.

Reply:
updated